### PR TITLE
Layer package proof of concept

### PR DESCRIPTION
**Pull request by dmcgowan**
Proof of concept for layer store interface

Note tests don't currently work due to issue with chroot archive, working on it

### CODE REVIEW

**Comment by tonistiigi** [comment]
For the cacheIDs to work we need to keep a file on disk with digest-cacheid mappings. I wonder if after this has been implemented the cacheLayer linkedlist makes much sense. I think it would be easier to keep simple layer object that has references to digest ID and digest ParentID and just provide a lookup to cacheID where its needed.

Also, why do you call it `cacheID`, I don't think it has any semantics of a cache. How about localID, generatedID, graphID, driverID.

File: layers/layers.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	"github.com/docker/docker/daemon/graphdriver"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/stringid"
+)
+
+var (
+	ErrLayerDoesNotExist = errors.New("layer does not exist")
+)
+
+type LayerAddress digest.Digest
```
Comment (tonistiigi) [inline layers/layers.go]: I think "address" is confusing, it should be ID or digest.

File: layers/layers.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+type TarStreamer interface {
+	TarStream() (io.Reader, error)
+}
+
+// Layer represents a read only layer
+type Layer interface {
+	TarStreamer
+	Address() LayerAddress
+	Parent() (Layer, error)
+	Size() (int64, error)
```
Comment (tonistiigi) [inline layers/layers.go]: Why interface and why not just properties of a defined struct(parent would be Parent digest).

File: layers/layers.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+	layer.size, err = ls.driver.ApplyDiff(layer.cacheID, pid, archive.Reader(tr))
+	if err != nil {
+		return nil, err
+	}
+
+	layer.tarStreamer = func() (io.Reader, error) {
+		archiver, err := ls.driver.Diff(layer.cacheID, pid)
+		return io.Reader(archiver), err
+	}
```
Comment (tonistiigi) [inline layers/layers.go]: Why do we need this as inline func.

File: layers/layers.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+		}
+		p = l
+		pid = l.cacheID
+	}
+
+	mount := &mountedLayer{
+		parent:  p,
+		mountID: stringid.GenerateRandomID(),
+	}
+
```
Comment (tonistiigi) [inline layers/layers.go]: Check exists first.

File: layers/layers.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	defer ls.mountL.Unlock()
+
+	m := ls.mounts[id]
+	if m == nil {
+		return errors.New("mount does not exist")
+	}
+
+	delete(ls.mounts, id)
+
+	return ls.driver.Remove(m.mountID)
```
Comment (tonistiigi) [inline layers/layers.go]: Plain unmount should be `driver.Put`. Do we need a separate way to handle the removing? Btw, I think currently its not safe to call mount/unmount multiple times, Michael is been trying to fix that for aufs https://github.com/docker/docker/pull/15868

**Comment by dmcgowan** [comment]
I wanted to use a name that indicates that the layer is immutable. However the word "cache" probably only makes sense if we have an artifact store which is the source of truth, cache implies the source is from somewhere else. `graphID` can make sense since it is what you would use for the `graphdriver`, although that name is a different story.

I agree on storing the mappings on disk, I am going to separate out that into a separate interface. Linked list is certainly not necessary, just preferred checking for `parent == nil` instead of `parentAddress == ""`, in the end would not change the code layout.

File: layers/layers.go (hunk 3 above)
Comment (dmcgowan) [inline layers/layers.go]: I think after our discussion it makes sense to move these back to to layer store object

**Pull request closed without merging**